src/allotropy/parsers/thermo_fisher_nanodrop_eight/nanodrop_eight_parser.py

```diff
@@ -0,0 +1,302 @@
+from allotropy.allotrope.models.spectrophotometry_benchling_2023_12_spectrophotometry import (
+    CalculatedDataAggregateDocument,
+    CalculatedDataDocumentItem,
+    DataSourceAggregateDocument1,
+    DataSourceDocumentItem,
+    DataSystemDocument,
+    DeviceSystemDocument,
+    MeasurementAggregateDocument,
+    Model,
+    SampleDocument,
+    SpectrophotometryAggregateDocument,
+    SpectrophotometryDocumentItem,
+    UltravioletAbsorbancePointDetectionDeviceControlAggregateDocument,
+    UltravioletAbsorbancePointDetectionDeviceControlDocumentItem,
+    UltravioletAbsorbancePointDetectionMeasurementDocumentItems,
+)
+from allotropy.constants import ASM_CONVERTER_NAME, ASM_CONVERTER_VERSION
+from allotropy.named_file_contents import NamedFileContents
+from allotropy.parsers.thermo_fisher_nanodrop_eight.nanodrop_eight_reader import (
+    NanoDropEightReader,
+)
+from allotropy.parsers.utils.values import assert_not_none
+from allotropy.parsers.vendor_parser import VendorParser
+
+ConcentrationType = Union[
+    TQuantityValueMicrogramPerMicroliter,
+    TQuantityValueMicrogramPerMilliliter,
+    TQuantityValueMilligramPerMilliliter,
+    TQuantityValueNanogramPerMicroliter,
+    TQuantityValueNanogramPerMilliliter,
+    TQuantityValuePicogramPerMilliliter,
+]
+ConcentrationClassType = Union[
+    type[TQuantityValueMicrogramPerMicroliter],
+    type[TQuantityValueMicrogramPerMilliliter],
+    type[TQuantityValueMilligramPerMilliliter],
+    type[TQuantityValueNanogramPerMicroliter],
+    type[TQuantityValueNanogramPerMilliliter],
+    type[TQuantityValuePicogramPerMilliliter],
+]
+
+CONCENTRATION_UNIT_TO_TQUANTITY: Mapping[str, ConcentrationClassType] = {
+    "ug/ul": TQuantityValueMicrogramPerMicroliter,
+    "ug/ml": TQuantityValueMicrogramPerMilliliter,
+    "mg/ml": TQuantityValueMilligramPerMilliliter,
+    "ng/ul": TQuantityValueNanogramPerMicroliter,
+    "ng/ml": TQuantityValueNanogramPerMilliliter,
+    "pg/ul": TQuantityValuePicogramPerMilliliter,
+}
+
+
+def _get_str_or_none(data_frame: pd.DataFrame, row: int, column: str) -> Optional[str]:
+    if column not in data_frame.columns:
+        return None
+
+    val = data_frame.iloc[row][column]
+    if pd.isna(val):
+        return None
+
+    return str(val)
+
+
+def _get_str(data_frame: pd.DataFrame, row: int, column: str) -> str:
+    val = _get_str_or_none(data_frame=data_frame, row=row, column=column)
+
+    assert_not_none(val)
+
+    return str(val)
+
+
+def _get_float(data_frame: pd.DataFrame, row: int, column: str) -> JsonFloat:
+    try:
+        return float(data_frame.iloc[row][column])
+    except (ValueError, TypeError):
+        return InvalidJsonFloat.NaN
+
+
+def _get_concentration(
+    conc: JsonFloat, unit: Optional[str]
+) -> Optional[ConcentrationType]:
+    if unit in CONCENTRATION_UNIT_TO_TQUANTITY and isinstance(conc, float):
+        cls = CONCENTRATION_UNIT_TO_TQUANTITY[str(unit)]
```

Comment:
nit: this `str(` call should not be necessary, if `unit` is in `CONCENTRATION_UNIT_TO_TQUANTITY`, the it must already be a string.

Reply:
I agree but the type checker is unhappy without the `str(` call.  Is there a better way?

Reply:
Oh of course it's the type checker 🙄 

I think if you add `if unit and unit in ...` to the front of the conditional it will be happy.

If that doesn't work, you can just leave it.